feat: add dashboard lightweight card

kertuilves opened the pull request:
https://app.clickup.com/t/861n0qm47

heshfekry commented:
@kertuilves just confirming that this is the component 

https://deploy-preview-289--conversionxl-aybolit.netlify.app/?path=/story/cxl-ui-cxl-card--cxl-course

heshfekry commented:
Hey @kertuilves 

Would be good to see multiple cards in the same view so we can assess how they work together? [Similar to this.](https://deploy-preview-289--conversionxl-aybolit.netlify.app/?path=/story/cxl-ui-cxl-vaadin-accordion--cxl-vaadin-accordion-theme-archive)


Right now it seems to take full width,

![image](https://github.com/conversionxl/aybolit/assets/72975440/d6a8b1d1-7538-43b2-8ad8-e5491af9ca55)

 so unclear how it would function with multiple cards
 
 
![image](https://github.com/conversionxl/aybolit/assets/72975440/4d16458b-8cce-4966-9ccd-35166de41f14)


Additionally
1.  would be good to see the indicator.
2.  A problem came up yday when we where working on roadmaps from teams and personal. [We probably need to have tags on there too indicating if the course has been selected in roadmap or not.](https://www.figma.com/file/RTj2PtUSrCOUFtljOiq0W3/Dashboard?node-id=445%3A39255&mode=dev)
- I would change the label from Team > Team Roadmap and from Personal > Personal Roadmap

![image](https://github.com/conversionxl/aybolit/assets/72975440/eb8385ba-832c-44ef-9234-155e15d47131)

pawelkmpt reviewed (changes requested):
Icon is cut

![Screenshot 2023-07-05 at 12 03 37](https://github.com/conversionxl/aybolit/assets/5165721/9cbf692a-0b52-4e36-99a0-ee40ece52f29)

pawelkmpt commented:
> A problem came up yday when we where working on roadmaps from teams and personal. [We probably need to have tags on there too indicating if the course has been selected in roadmap or not.](https://www.figma.com/file/RTj2PtUSrCOUFtljOiq0W3/Dashboard?node-id=445%3A39255&mode=dev)

Use vaadin badge for tags.

kertuilves commented:
Hey @heshfekry.
1. I modified the Course Card canvas layout and now the icon is [visible](https://deploy-preview-289--conversionxl-aybolit.netlify.app/?path=/story/cxl-ui-cxl-card--cxl-course&args=State:new)
2. Course Card now has another `with-roadmaps` [state](https://deploy-preview-289--conversionxl-aybolit.netlify.app/?path=/story/cxl-ui-cxl-card--cxl-course&args=State:new)
3. I added an example under [cxl-tabs-slider](https://deploy-preview-289--conversionxl-aybolit.netlify.app/?path=/story/cxl-ui-cxl-tabs-slider--cxl-course-slider) to see multiple Course Cards in the same view

heshfekry commented:
> Hey @heshfekry.
> 
> 1. I modified the Course Card canvas layout and now the icon is [visible](https://deploy-preview-289--conversionxl-aybolit.netlify.app/?path=/story/cxl-ui-cxl-card--cxl-course&args=State:new)
> 2. Course Card now has another `with-roadmaps` [state](https://deploy-preview-289--conversionxl-aybolit.netlify.app/?path=/story/cxl-ui-cxl-card--cxl-course&args=State:new)
> 3. I added an example under [cxl-tabs-slider](https://deploy-preview-289--conversionxl-aybolit.netlify.app/?path=/story/cxl-ui-cxl-tabs-slider--cxl-course-slider) to see multiple Course Cards in the same view

Looking good. The size seems to change with different states, unclear how that plays with multiple cards in the same view.

kertuilves commented:
@heshfekry 

> Looking good. The size seems to change with different states, unclear how that plays with multiple cards in the same view.

I kept the card itself really basic and didn't constrain its component width. In my opinion this should be done when adding the card to the specific layout. In our example when adding it to the slider in the accordion. And those slider styles and card widths are right now visible in the [accordion](https://deploy-preview-288--conversionxl-aybolit.netlify.app/?path=/story/cxl-ui-cxl-vaadin-accordion--cxl-vaadin-accordion-theme-category) PR.

heshfekry commented:
> @heshfekry
> 
> > Looking good. The size seems to change with different states, unclear how that plays with multiple cards in the same view.
> 
> I kept the card itself really basic and didn't constrain its component width. In my opinion this should be done when adding the card to the specific layout. In our example when adding it to the slider in the accordion. And those slider styles and card widths are right now visible in the [accordion](https://deploy-preview-288--conversionxl-aybolit.netlify.app/?path=/story/cxl-ui-cxl-vaadin-accordion--cxl-vaadin-accordion-theme-category) PR.

This makes sense. Does this apply to height too? for example when you introduce the tags or the names etc take more space.

kertuilves commented:
> > @heshfekry
> > > Looking good. The size seems to change with different states, unclear how that plays with multiple cards in the same view.
> > 
> > 
> > I kept the card itself really basic and didn't constrain its component width. In my opinion this should be done when adding the card to the specific layout. In our example when adding it to the slider in the accordion. And those slider styles and card widths are right now visible in the [accordion](https://deploy-preview-288--conversionxl-aybolit.netlify.app/?path=/story/cxl-ui-cxl-vaadin-accordion--cxl-vaadin-accordion-theme-category) PR.
> 
> This makes sense. Does this apply to height too? for example when you introduce the tags or the names etc take more space.

Yes. So all the cards heights will be the same and it depends on the highest card. I kept it the same way as it is for the slider cards by default. 

<img width="1533" alt="Screenshot 2023-07-05 at 15 19 00" src="https://github.com/conversionxl/aybolit/assets/119847199/43401edc-7bf1-48b2-b614-1d3c41203c70">

heshfekry commented:
@freudFlintstone pulling you in here because you are working on upgrading the full card.

You mentioned that this could all be the same component with an attribute. Does it make sense you take over all the card components?

heshfekry commented:
@freudFlintstone lets continue as is with now as separate components ill make another CU card to revisit the combination project at the end once we have everything in place.

pawelkmpt commented on packages/cxl-ui/scss/global/cxl-card.scss:
```diff
@@ -24,4 +24,58 @@ cxl-card {
+        right: 0;
+        width: calc(var(--lumo-space-m) * 2);
+        height: calc(var(--lumo-space-m) * 2);
+        padding: 6px;
```
I meant just `padding` ;) but as long as we don't have more icons it's fine

pawelkmpt commented on packages/storybook/cxl-ui/cxl-card/[theme=cxl-course].stories.js:
```diff
@@ -0,0 +1,67 @@
+CXLCourse.argTypes = {
+  State: {
+    control: {
+      options: ['default', 'new', 'completed', 'with-roadmaps'],
```
I suggest to change `with-roadmaps` to `with-badges` to make it more universal as badges might be used also for tags etc.

And question: is it possible to apply multiple states at the same time?

kertuilves replied:
Changed the roadmaps -> badges and now it's possible to select multiple [states](https://deploy-preview-289--conversionxl-aybolit.netlify.app/?path=/docs/cxl-ui-cxl-card--cxl-course)

pawelkmpt commented on packages/storybook/cxl-ui/cxl-card/[theme=cxl-course].stories.js:
```diff
@@ -0,0 +1,77 @@
+      theme="cxl-course ${stateArray.map(state => `cxl-course-${state}`).join(' ')}"
+    >
+      <header class="entry-header">
+        ${stateArray.map(state =>
+          state === 'new'
+          ? html`<vaadin-icon icon="cxl:new"></vaadin-icon>`
+          : html``
+        )}
+        <div>
+          <div class="entry-title">
+            Account based marketing
+            ${stateArray.map(state =>
+              state === 'completed'
+              ? html`<vaadin-icon icon="lumo:checkmark"></vaadin-icon>`
+              : html``
+            )}
+          </div>
+          <div class="entry-byline">
+            <div><vaadin-icon icon="lumo:clock"></vaadin-icon>3h 00min</div>
+            <div>By: Ton Wesseling</div>
+          </div>
+          ${stateArray.map(state =>
+            state === 'with-badges'
+            ? html`
+              <div class="entry-summary">
+                <span theme="badge primary pill">Team Roadmap</span>
+                <span theme="badge secondary pill">Personal Roadmap</span>
+              </div>`
+            : html``
+          )}
+        </div>
```
I'm concerned about the logic being present in storybook instead of encapsulating it into the component.

As I see, adding state param eg. `cxl-course-new` will not make new icon visible. It'll just apply specific styles, but if icon is not present... card will look the same as without this state.

For completed the same: we need to manually put an icon instead of just adding `cxl-course-completed` theme element.

Which makes me think: why do we even need to define states?

@lkraav what's your opinion?

pawelkmpt replied:
With badges it should just work out of the box: put badges and they're displayed nicely without a need to add theme name.

lkraav replied:
> As I see, adding state param eg. `cxl-course-new` will not make new icon visible. It'll just apply specific styles, but if icon is not present... card will look the same as without this state.
> 
> For completed the same: we need to manually put an icon instead of just adding `cxl-course-completed` theme element.
> 
> Which makes me think: why do we even need to define states?

I think I agree - if possible, we should aim for an atomic (single) operation to achieve our goal. "Need to know multiple steps" is road to confusion.

pawelkmpt replied:
Alright, @kertuilves please encapsulate these states in the component.

`new` -> icon appears (without a need to put icon manually)

`completed` -> checkmark (without a need to put icon manually)

`with-badges` -> drop state; badges need to look good without need to add additional theme, just `cxl-course`

pawelkmpt reviewed (commented):
Looking good. Concerns about logic.

anoblet commented on packages/cxl-lumo-styles/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@conversionxl/cxl-lumo-styles",
-  "version": "1.6.3",
+  "version": "1.6.4",
```
I'm not sure it's best practice to bump package versions with each PR. If so, it would be better in a separate commit.

@pawelkmpt @lkraav Thoughts?

pawelkmpt commented:
@kertuilves it seems like you squashed your changes with existing commit during rebase. Please correct it. Are you familiar with `git reflog`?

pawelkmpt commented:
@kertuilves you pushed wrong code here `cxl-stats`

kertuilves commented:
> @kertuilves you pushed wrong code here `cxl-stats`

@pawelkmpt The `cxl-stats` has been merged to feat/dashboard and when I rebased this branch from feat/dashboard and solved the merge conflicts it was supposed to come. No?

pawelkmpt commented:
> @pawelkmpt The `cxl-stats` has been merged to feat/dashboard and when I rebased this branch from feat/dashboard and solved the merge conflicts it was supposed to come. No?

Everything is fine. My bad. I looked at _Compare_ option https://github.com/conversionxl/aybolit/compare/9f2bfc79dfe1adc13fb54768514c200119990f29..c9487b473530c54471bd36a12314245061c9e8c6, shown stats component there and I had an impression it's pushed here.

Sorry for that

pawelkmpt commented:
@kertuilves taking into account Slack threads about the way of developing components:
* https://cxlworld.slack.com/archives/C01HXUNGEPM/p1689060114569099
* https://cxlworld.slack.com/archives/C01HXUNGEPM/p1689084292805349

I think we need change in this code. @freudFlintstone has almost completed `<cxl-course-card>` and you can look at his code here: https://github.com/conversionxl/aybolit/commit/3e46d1286e2d7c4a8b457581a4b58b42be0f679d#diff-5e748d1f0301dffd4bb4521b9dbe089c8a76e34023e06745e8848cd542447a6e

Main difference: component has set of attributes and properties which are encapsulated and rendered in specific way.

Good thing is that you can re-use styles you already created :)

What's next:
* keep `cxl-card` as is,
* created new component `cxl-light-card`.

kertuilves commented:
@pawelkmpt I created a separate component for light-card. How should I pass the badges data? Do we know how will the different roadmaps themes be defined? (will work on it after vacation)

pawelkmpt commented on packages/cxl-ui/src/components/cxl-light-card.js:
```diff
@@ -0,0 +1,51 @@
+
+  @property({ type: String }) theme = 'light-card';
+
+  @property({ type: String }) title = 'Account based marketing';
+
+  @property({ type: String }) time = '3h 00min';
```
Default should be empty string

pawelkmpt commented on packages/cxl-ui/src/components/cxl-light-card.js:
```diff
@@ -0,0 +1,51 @@
+
+  @property({ type: String }) instructorPrefix = 'By:';
+
+  @property({ type: String }) instructor = 'Ton Wesseling';
```
Default should be empty string

pawelkmpt commented on packages/cxl-ui/src/components/cxl-light-card.js:
```diff
@@ -0,0 +1,51 @@
+
+  @property({ type: String }) instructor = 'Ton Wesseling';
+
+  @property({ type: String }) avatar =
```
Default should be empty string

pawelkmpt commented on packages/cxl-ui/src/components/cxl-light-card.js:
```diff
@@ -0,0 +1,51 @@
+              <div><vaadin-icon icon="lumo:clock"></vaadin-icon>${this.time}</div>
+              <div>${this.instructorPrefix}&nbsp;${this.instructor}</div>
+            </div>
+            <slot name="badges"></slot>
```
> How should I pass the badges data? Do we know how will the different roadmaps themes be defined?

I think we should rename it to `footer` and let us pass any HTML we want. It will be more flexible and easier to manage. Today requirement is for badges but it might be also text tags etc.

pawelkmpt commented on packages/cxl-ui/src/components/cxl-light-card.js:
```diff
@@ -0,0 +1,51 @@
+import cxlLightCardStyles from '../styles/cxl-light-card-css.js';
+
+@customElement('cxl-light-card')
+export class CXLCourseCardElement extends LitElement {
```
I think this should have a different name

pawelkmpt commented on packages/cxl-ui/src/components/cxl-light-card.js:
```diff
@@ -0,0 +1,50 @@
+
+  @property({ type: String }) avatar = '';
+
+  @property({ type: Boolean }) hasFooter = false;
```
Is it needed to display footer? Can't we just put content into slot and render it automatically and it's there's nothing to display, card lives without footer?

pawelkmpt commented on packages/cxl-ui/src/components/cxl-light-card.js:
```diff
@@ -0,0 +1,50 @@
+    return html`
+      <div class="container">
+        <header class="entry-header">
+          <vaadin-icon icon="cxl:new"></vaadin-icon>
```
"New" is used as a property in #294 and #293 and we should make it same here

kertuilves replied:
I made the "new" as a property and it made sense to use the same pattern with completed card. So now the "completed" is also a separate property and I removed the themes because these are not necessary anymore.

pawelkmpt commented on packages/cxl-lumo-styles/scss/badge.scss:
```diff
@@ -7,6 +7,11 @@
   color: var(--lumo-primary-text-color);
 }
 
+[theme~='badge'][theme~='secondary'] {
+  color: var(--lumo-primary-contrast-color);
+  background-color: var(--cxl-color-brand-blue);
+}
+
```
Please make it separate commit as it's badge feature

pawelkmpt replied:
@kertuilves following it up

pawelkmpt reviewed (changes requested):
Navigation differs from Figma design

it is:
<img width="433" alt="Screenshot 2023-07-27 at 15 57 23" src="https://github.com/conversionxl/aybolit/assets/5165721/7c48cc04-5b3d-4922-b9b1-1093e71dc08e">

should be:
<img width="291" alt="Screenshot 2023-07-27 at 15 57 08" src="https://github.com/conversionxl/aybolit/assets/5165721/afdc1dfd-af85-4e8f-99b3-1bdfcdf77679">

kertuilves commented:
> Navigation differs from Figma design

I made the slider changes now here, too. But this was already discussed above that those slider styles were already under #288. And the slider here was just to show how multiple cards together would behave.

pawelkmpt commented on packages/cxl-lumo-styles/src/icons.js:
```diff
@@ -53,6 +53,9 @@ $documentContainer.innerHTML = `
           <path d="M5.614.416a4.538 4.538 0 011.824.086A4.259 4.259 0 005.614.416zM2.322 12.449l.002.009c-.747-1.19-.94-.39-.97-.224-.587-1.581-.96-2.743-.96-2.743a10.93 10.93 0 01-.15-.61L.22 8.77c-.034-.168-.063-.33-.09-.484v-.003a16.671 16.671 0 01-.077-.583C-.04 6.725.04 5.89.23 5.176c.011-.039.02-.08.034-.117.01-.034.02-.07.033-.102.335-1.039.72-1.273.72-1.273 1-.892.963.434.963.434-1.05 2.3 ...
           <path d="M1.355 12.234l-.006.029A17.108 17.108 0 01.243 8.88c.043.2.094.403.15.61 0 0 .374 1.16.962 2.743zM.054 7.7C-.064 6.576.05 5.967.05 5.967c.055-.3.116-.563.18-.79-.19.714-.269 1.549-.175 2.525z"/>
         </g>
+        <g id="new" width="20" height="20" viewBox="0 0 20 20">
```
I think this branch needs rebase on top of `feat/dashboard`. the other branch already have "new" icon https://github.com/conversionxl/aybolit/blob/e4a46aa4647d7882365e5a79aaf7b08ce7481d93/packages/cxl-lumo-styles/src/icons.js#L56-L58

pawelkmpt reviewed (changes requested):
Please put badge code into one commit and rest into another

pawelkmpt commented on packages/cxl-lumo-styles/scss/global.scss:
```diff
@@ -8,6 +8,7 @@ html {
   --cxl-wrap-padding: var(--lumo-space-m);
   --cxl-color-light-gray: hsla(0, 0%, 96%, 1);
   --cxl-color-black-50pct: hsla(0, 0%, 0%, 0.5);
+  --cxl-color-brand-blue: hsla(214, 61%, 25%, 1);
```
This color should be in the other branch

pawelkmpt reviewed (approved):
Good job 👍 
Next step: refactor to use base class

pawelkmpt commented on packages/cxl-ui/src/components/cxl-base-card.js:
```diff
@@ -50,16 +54,19 @@ export class CXLBaseCardElement extends LitElement {
+          ? html`<vaadin-icon icon="lumo:checkmark"></vaadin-icon>`
+          : nothing}
+      </div>
+    `;
   }
 
   _renderHeaderAttributes() {
     return html`
       <div class="attributes">
-        <cxl-time
-          time=${this.time}
-          ?show-icon=${!!(this.theme.toLowerCase() === 'course')}
-        ></cxl-time>
+        <cxl-time time=${this.time} ?show-icon=${this.showTimeIcon}></cxl-time>
```
Changing this logic impacts `<cxl-course-card>`. You set it ` @state() showTimeIcon = true;` in the component but it's different then it was originally. Please revisit original code and adjust

pawelkmpt commented on packages/cxl-ui/scss/cxl-base-card.scss:
```diff
@@ -157,8 +157,8 @@ header {
     position: absolute;
-    top: calc(-1 * var(--lumo-space-s));
-    right: calc(-1 * var(--lumo-space-s));
+    top: calc(-1 * var(--lumo-space-m));
+    right: 0;
```
I see there's difference in Figma design between different cards. I experimented with values and I think original style here makes more sense, so please revert.

pawelkmpt commented on packages/cxl-ui/scss/cxl-time.scss:
```diff
@@ -1,5 +1,6 @@
 :host {
   display: flex;
+  align-items: center;
```
👍

pawelkmpt commented on packages/cxl-ui/scss/cxl-base-card.scss:
```diff
@@ -157,8 +157,8 @@ header {
-    top: calc(-1 * var(--lumo-space-m));
-    right: 0;
+    top: calc(-1 * var(--lumo-space-s));
+    right: calc(-1 * var(--lumo-space-s));
```
This should go into 404b91d37d4eb09fc60ca27f4c3df7a2c2c8da3c

pawelkmpt commented on packages/cxl-ui/src/components/cxl-base-card.js:
```diff
@@ -33,7 +33,7 @@ export class CXLBaseCardElement extends LitElement {
   @property({ type: Boolean, reflect: true }) completed = false;
 
-  @property({ type: Boolean, reflect: true }) showTimeIcon = false;
+  @property({ type: Boolean, reflect: true, attribute: 'show-time-icon' }) showTimeIcon = false;
```
This should go into 404b91d37d4eb09fc60ca27f4c3df7a2c2c8da3c

pawelkmpt commented on packages/cxl-ui/src/components/cxl-course-card.js:
```diff
@@ -14,8 +14,6 @@ export class CXLCourseCardElement extends CXLBaseCardElement {
   @state() _moreHasChildren = false;
 
-  @state() showTimeIcon = true;
-
```
This should go into 404b91d37d4eb09fc60ca27f4c3df7a2c2c8da3c

pawelkmpt commented on packages/storybook/cxl-ui/cxl-course-card/[theme=playbook].stories.js:
```diff
@@ -1,23 +1,24 @@
   contentTags: [],
-  tags: ['Growth Marketing']
+  tags: ['Growth Marketing'],
+  showTimeIcon: false,
```
This should go into 404b91d37d4eb09fc60ca27f4c3df7a2c2c8da3c

pawelkmpt commented on packages/storybook/cxl-ui/cxl-course-card/[theme=training].stories.js:
```diff
@@ -1,23 +1,24 @@
   contentTags: [],
-  more: ''
+  more: '',
+  showTimeIcon: false,
```
This should go into 404b91d37d4eb09fc60ca27f4c3df7a2c2c8da3c

pawelkmpt commented on packages/storybook/cxl-ui/cxl-course-card/[theme=video].stories.js:
```diff
@@ -1,24 +1,25 @@
   time: '4 days ago',
-  new: true
+  new: true,
+  showTimeIcon: false,
```
This should go into 404b91d37d4eb09fc60ca27f4c3df7a2c2c8da3c

pawelkmpt commented on packages/storybook/cxl-ui/cxl-course-card/template.js:
```diff
@@ -1,38 +1,51 @@
 import { html } from 'lit';
 
-const renderTags = (tags, slot) => tags.map((tag, i) => html`${i > 0 ? html`<span slot=${slot}> | </span>` : ''}<span class="tag" slot=${slot} title="${tag}">${tag}</span>`);
+const renderTags = (tags, slot) =>
```
This should go into 404b91d37d4eb09fc60ca27f4c3df7a2c2c8da3c

pawelkmpt reviewed (changes requested):
@kertuilves I marked couple of changes which should go into "improve base card" commit. For storybook templates it's not just a line I left the comment but all changes which are caused by `showTimeIcon` property introduction.

Light card itself looks good
